### ISSUE TITLE
No way to apply GOGS_CUSTOM for SSH

### ISSUE THREAD

**Issue by klausenbusk** [gogits/gogs#792]
- **Bug Description**: 
There seems to be no way to apply GOGS_CUSTOM for SSH, which is making GOGS_CUSTOM nearly useless, as you can't get SSH to use the same config..
- **Gogs Version**: `v0.5.9.1216 Beta`
- **Git Version**: `2.2.1`
- **System Type**: `Arch Linux`
- **Error Log**: Nothing
- **Other information**: #486 have some useful information (bottom)

**Comment by hickscorp** [comment]
**A solution for running multiple Gogs daemons with the same unix user would probably be a PAM module, but i am not that familiar with it, so please correct me if i’m wrong. For instance, looking at /etc/pam.d/ssh shows that one of the first strategy to be loaded when authenticating SSH is pam_loginuid.so. Putting a custom compilable module before this step would make this possible, but isn’t it really too much for what we want? IMO, having N users for N instances is very clean, and is the “unix” way to do things.**

The binary argument seems the best option to me too.
As @phsmit  stated, there could be a number of possible arguments to *override*.
By default, Gogs could be looking with priority:
- Hard-coded defaults when initializing the configuration structures,
- Read defaults from /etc/gogs if they exist,
- Read overrides (Or full config if all parameters are in the file) from a location pointed by -config “/my/custom”
Also as pointed by @phsmit, this could also work with the templates, and the variable data (Logs / PIDs). System defaults overridable by executable flag.

**Comment by Unknwon** [comment]
> Instead, why not search for the current directory (The directory where the user running the daemon was positioned when starting Gogs)?

This way we can apply your custom config when `./gogs serve` been called when you push.

**Comment by hickscorp** [comment]
@Unknwon Yes :) But again, @phsmit pointed a more reliable solution IMO, the flags parsing option.

Please see my edited post for more soutions about the SSH keys for multiple instances.

**Comment by Unknwon** [comment]
> I suggest making the configlocation an argument of gogs web, with as default the current situation as default. The same for a sharelocation (where the templates are stored). If you like I can either give a detailed description or implement it in pull request and discuss that further ( @Unknwon which do you prefer?)

This is problem we have right now, `./gogs serve` is called by system(actually `git` itself), the argument you pass into gogs web has no way to tell when `gogs serve` is executed.

**Comment by Unknwon** [comment]
PS:

Unless every time you run Gogs with option(where is the custom path), Gogs rewrites entire authorized_keys file.

**Comment by hickscorp** [comment]
@Unknwon Understood. I really think multi-instances installs can stay with as many unix users right now. But it's worth investigation, either with the UUID system or with the PAM module whenever people really want this feature.

**Comment by hickscorp** [comment]
One bummer, use case:
- Two gogs running in the same user.
- One user being registered with the same SSH key in the two instances.

Right now, stick with the user-separation :)

**Comment by Unknwon** [comment]
@hickscorp 

I would prefer one user one instance.

But a thought on one user multiple instances, Gogs can do a SSH forward, which handles AUTH on Gogs, and forward operations to system(real) SSH server. This way, we do not need to care about default authorized_keys file any more, but just Gogs database check.

**Comment by klausenbusk** [comment]
Any update? I need to edit authorized_keys manually, for every new ssh key..

**Comment by Unknwon** [comment]
Hi @klausenbusk , why would this issue be matter with "I need to edit authorized_keys manually, for every new ssh key"? Sounds like it's about fail to add SSH key?

**Comment by klausenbusk** [comment]
> Hi @klausenbusk , why would this issue be matter with "I need to edit authorized_keys manually, for every new ssh key"? Sounds like it's about fail to add SSH key?

I need to change it from:
`command="/usr/share/gogs/gogs serv key-5"`
to:
`command="GOGS_CUSTOM="/srv/gogs" /usr/share/gogs/gogs serv key-5"`

So it can find the config file..

**Comment by Unknwon** [comment]
An update on:

`dev` branch only saves partial configuration when install, so it's possible to write config options(that not conflict with) in other places.

**Comment by Unknwon** [comment]
https://github.com/gogits/gogs/pull/905 this PR will help on this

**Comment by Unknwon** [comment]
By PR https://github.com/gogits/gogs/pull/905 and other code I've pushed to `dev` branch, you are now able to fix this issue.

PS: you can now also do rewrite SSH key file and all repository update hooks in admin dashboard with one click.

**Comment by klausenbusk** [comment]
@Unknwon I'm trying to compile the dev branch, but I get this error
```
WORK=/tmp/go-build952790733
../../macaron-contrib/cache/memcache/memcache.go:22:2: cannot find package "github.com/bradfitz/gomemcache/memcache" in any of:
	/mnt/gogs/src/build/go/src/github.com/bradfitz/gomemcache/memcache (from $GOROOT)
	/mnt/gogs/src/build/src/github.com/bradfitz/gomemcache/memcache (from $GOPATH)
../../microcosm-cc/bluemonday/sanitize.go:38:2: cannot find package "golang.org/x/net/html" in any of:
	/mnt/gogs/src/build/go/src/golang.org/x/net/html (from $GOROOT)
	/mnt/gogs/src/build/src/golang.org/x/net/html (from $GOPATH)
modules/base/template.go:16:2: cannot find package "golang.org/x/net/html/charset" in any of:
	/mnt/gogs/src/build/go/src/golang.org/x/net/html/charset (from $GOROOT)
	/mnt/gogs/src/build/src/golang.org/x/net/html/charset (from $GOPATH)
modules/base/template.go:17:2: cannot find package "golang.org/x/text/transform" in any of:
	/mnt/gogs/src/build/go/src/golang.org/x/text/transform (from $GOROOT)
	/mnt/gogs/src/build/src/golang.org/x/text/transform (from $GOPATH)
```
Don't know if it is something in your code or.. Not a Go expert :)

**Comment by Unknwon** [comment]
@klausenbusk this means you need to download some of packages that are newly added in `dev` branch. 

Following commands should help:

```shell
go get github.com/bradfitz/gomemcache/memcache
go get golang.org/x/net/
go get golang.org/x/text
```

**Comment by klausenbusk** [comment]
> Following commands should help:

I will try.. Shouldn't they be in .gopmfile or is that for "other stuff"?

**Comment by Unknwon** [comment]
Hmm... yeah, gopmfile updates before every release. Are you using gopm right? Or just use the gopmfile?

**Comment by klausenbusk** [comment]
> Hmm... yeah, gopmfile updates before every release. Are you using gopm right? Or just use the gopmfile?

At the momemt, I'm just patching the ArchLinux PKGBUILD file.. Which seems to parse the .gopmfile and use go get  : https://github.com/fanningert/PKGBUILDs/blob/master/aur/gogs/helper.sh#L52

**Comment by Unknwon** [comment]
I see, I've pushed fix on `dev`, please help try again!

**Comment by klausenbusk** [comment]
> I see, I've pushed fix on dev, please help try again!

```
package golang.org/x/net
	imports golang.org/x/net
	imports golang.org/x/net: no buildable Go source files in /mnt/gogs/src/build/src/golang.org/x/net
```
also got that error, when I tried go get in the PKGBUILD..

**Comment by Unknwon** [comment]
It's a warning, you can just ignore it when download is good

**Comment by klausenbusk** [comment]
> It's a warning, you can just ignore it when download is good

I did `go_get ${array[0]} "$2/${array[0]}" master || true`, there don't seems to be a flag for it?
New compile error http://dpaste.com/14W5J1B (5 last line)

**Comment by Unknwon** [comment]
Fine... I'm nor familiar with that, how about test this on newer release which comes soon?

**Comment by klausenbusk** [comment]
> Fine... I'm nor familiar with that, how about test this on newer release which comes soon?

okay :) I don't have any problems waiting, there is only 2 users on my gogs "instance", so I can handle it manually..

**Comment by Unknwon** [comment]
@klausenbusk sorry for that, this issue actually improves a lot things in Gogs.

**Comment by klausenbusk** [comment]
> @klausenbusk sorry for that, this issue actually improves a lot things in Gogs.

I think I somehow have fucked something up, while testing.. http://dpaste.com/3KX9HCP Maybe the database need to be downgraded? I did recompile 0.5.11, after I was finish testing..

**Comment by Unknwon** [comment]
Try update all dependencies,,,

**Comment by klausenbusk** [comment]
> Try update all dependencies,,,

You mean everything in gopmfile? I fetch that everytime it compile.. 
It still don't work, just did a recompile now (from a clean "build" environment)..

**Comment by Unknwon** [comment]
Just do `go get -u` on the root directory of source code of Gogs.

**Comment by klausenbusk** [comment]
> Just do go get -u on the root directory of source code of Gogs.

So I should switch to master?

**Comment by Unknwon** [comment]
I think stupid Go 1.4 will force change it to `master`... If I remember correctly.

**Comment by klausenbusk** [comment]
> I think stupid Go 1.4 will force change it to master... If I remember correctly.

At the momemt, I just fetch it with git and the 0.5.11 tag

**Comment by klausenbusk** [comment]
I think I just wait for a release (hopeful not too long), master and dev branch won't compile (http://dpaste.com/2KXNEM1)

**Comment by Unknwon** [comment]
Sure, planned this week.

**Comment by klausenbusk** [comment]
Fixed with #905.. Kind of..

**Comment by Unknwon** [comment]
:beer::beer::beer:
